File: COA software/backend/app/schemas/risk.py
from datetime import datetime
from typing import Optional, List, Dict, Any, Union
from pydantic import BaseModel, Field, field_validator


class RiskPredictionRequest(BaseModel):
    asset_id: str = Field(..., description="Asset UUID or Asset Code (e.g. TRK-1002)")
    horizon_days: int = Field(30, description="Prediction horizon in days (7, 30, 60, 90)")

    @field_validator("horizon_days")
    @classmethod
    def validate_horizon(cls, v: int) -> int:
        if v not in [7, 30, 60, 90]:
            raise ValueError("horizon_days must be one of [7, 30, 60, 90]")
        return v


class BulkRiskPredictionRequest(BaseModel):
    asset_ids: List[str] = Field(..., min_length=1, description="List of asset IDs or codes")
    horizon_days: int = Field(30, description="Prediction horizon in days (7, 30, 60, 90)")

    @field_validator("horizon_days")
    @classmethod
    def validate_horizon(cls, v: int) -> int:
        if v not in [7, 30, 60, 90]:
            raise ValueError("horizon_days must be one of [7, 30, 60, 90]")
        return v


class RiskFactor(BaseModel):
    factor: str
    raw_value: Union[float, int, str]
    normalized_value: float
    weight: float
    contribution: float
    severity: str


class RiskPredictionData(BaseModel):
    asset_id: str
    asset_code: str
    asset_name: Optional[str] = None
    asset_type: Optional[str] = None
    department: Optional[str] = None
    corridor_id: Optional[str] = None
    horizon_days: int
    risk_score: float
    risk_level: str
    failure_probability: float
    model: str
    model_version: Optional[str] = None
    recommendation: str
    explanation: Optional[str] = None
    factors: List[RiskFactor] = []
    prediction_date: Optional[str] = None


class RiskPredictionResponse(BaseModel):
    success: bool = True
    data: RiskPredictionData
    message: str = "Risk prediction generated successfully"


class BulkRiskPredictionResponse(BaseModel):
    success: bool = True
    data: Dict[str, Any]
    message: str = "Bulk risk predictions generated successfully"


class RiskHistoryItem(BaseModel):
    id: str
    asset_id: str
    prediction_date: datetime
    horizon_days: int
    risk_score: float
    risk_level: str
    failure_probability: float
    model_name: str
    model_version: str
    recommendation: Optional[str] = None
    explanation: Optional[str] = None

    class Config:
        from_attributes = True


class HighRiskAssetItem(BaseModel):
    id: str
    asset_id: str
    asset_code: str
    asset_name: str
    asset_type: str
    department: str
    corridor_id: Optional[str] = None
    health_score: float
    criticality_score: float
    risk_score: float
    risk_level: str
    failure_probability: float
    horizon_days: int
    recommendation: Optional[str] = None
    explanation: Optional[str] = None
    factors: List[RiskFactor] = []
    prediction_date: Optional[str] = None


class RiskSummaryData(BaseModel):
    critical_risk_count: int
    high_risk_count: int
    medium_risk_count: int
    low_risk_count: int
    total_predictions_monitored: int
    total_assets_count: int
    average_risk_score: float
    department_distribution: Dict[str, Dict[str, int]]
    risk_distribution: Dict[str, int]


class RiskSummaryResponse(BaseModel):
    success: bool = True
    data: RiskSummaryData
    message: str = "Risk summary retrieved successfully"

File: COA software/frontend/src/types/risk.ts
export interface RiskFactor {
  factor: string
  raw_value: number | string
  normalized_value: number
  weight: number
  contribution: number
  severity: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL'
}

export interface RiskPrediction {
  id?: string
  asset_id: string
  asset_code: string
  asset_name?: string
  asset_type?: string
  department?: string
  corridor_id?: string
  horizon_days: number
  risk_score: number
  risk_level: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL'
  failure_probability: number
  model: string
  model_version?: string
  recommendation: string
  explanation?: string
  factors: RiskFactor[]
  prediction_date?: string
}

export interface RiskHistoryItem {
  id: string
  asset_id: string
  prediction_date: string
  horizon_days: number
  risk_score: number
  risk_level: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL'
  failure_probability: number
  model_name: string
  model_version: string
  recommendation?: string
  explanation?: string
  factors?: RiskFactor[]
}

export interface HighRiskAsset {
  id: string
  asset_id: string
  asset_code: string
  asset_name: string
  asset_type: string
  department: string
  corridor_id?: string
  health_score: number
  criticality_score: number
  risk_score: number
  risk_level: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL'
  failure_probability: number
  horizon_days: number
  recommendation?: string
  explanation?: string
  factors: RiskFactor[]
  prediction_date?: string
}

export interface RiskSummary {
  critical_risk_count: number
  high_risk_count: number
  medium_risk_count: number
  low_risk_count: number
  total_predictions_monitored: number
  total_assets_count: number
  average_risk_score: number
  department_distribution: Record<string, Record<string, number>>
  risk_distribution: Record<string, number>
}
